### PR TITLE
Move NameRef and SymbolRef kind bits to end for better varint packing.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Move NameRef and SymbolRef kind bits to end for better varint packing.

### Motivation

We serialize `u4`s using variable-length integers (varints) to reduce the size of the payload baked into Sorbet and stored on disk in the cache. Varints are really nice for small numbers (numbers up to 2^7-1 fit in 1 byte!) but have overhead for large numbers (28+ bits require 5 bytes since it stores 7 bits of data per byte).

We store SymbolRef/NameRefs as u4s in the cache using their rawId, which encodes their kind and their index in that kind's table. Storing the kind bits in the _upper_ bits of rawId ensures that ~all of these references are stored inefficiently in the cache (5 bytes instead of 1 for small IDs!). I hadn't thought of this downside when initially engineering them this way. 

A secondary consequence of this inefficiency is that it's _slower_ to deserialize and serialize an integer from 5 bytes rather than 1 or 2. I am very interested in optimizing deserialization for my compressed AST work, which is the primary motivation of this work.

Fortunately, the solution is simple: Move the kind bits to the _end_ of the rawId. This solution shrinks payload by ~4% (0.1MB) and Stripe's cache directory (which contains ASTs) by ~2%. I do not see any noticeable changes in end-to-end runtime.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_IyHuV66Tgiv87k
→ https://go/builds/bui_IyHuTbGRmte7Gd

File: core/SymbolRef.h
Comment (froydnj) [inline core/SymbolRef.h:111]: Please change this comment to align with the code.

File: test/cli/simple-package/simple-package.out
Comment (froydnj) [inline test/cli/simple-package/simple-package.out:15]: These changes worry me.  Are we sorting on the raw id somewhere, and reordering the bits has changed things?  Is something reaching into the innards of `NameRef`/`SymbolRef` and didn't get updated for this change?  Or something else?

File: core/NameRef.h
Comment (froydnj) [inline core/NameRef.h:75]: Thank you for fixing up the "mask is max" nomenclature here.

**Review by froydnj — APPROVED** [review]
LGTM, but I would really like it if the test expectation changes were some combination of investigated/understood/fixed before this landed.

File: core/SymbolRef.h
Comment (jvilk-stripe) [reply core/SymbolRef.h:111]: Can do.

File: test/cli/simple-package/simple-package.out
Comment (jvilk-stripe) [reply test/cli/simple-package/simple-package.out:15]: Yes, `findMemberFuzzyMatchConstant` sorts on rawId:

https://github.com/sorbet/sorbet/blob/03f5d54c/core/Symbols.cc#L569

That's what is used for `Did you mean:` suggestions.

Comment (jvilk-stripe) [reply test/cli/simple-package/simple-package.out:15]: (The sorting is done for determinism reasons, not for semantic reasons.)

Comment (froydnj) [reply test/cli/simple-package/simple-package.out:15]: Works for me, thanks for clearing this up!

**Pull request merged**
